Mathlib/Tactic/CategoryTheory/ToApp.lean

```diff
@@ -0,0 +1,149 @@
+`Cat.whiskerLeft_app`, `Cat.whiskerRight_app`, `Cat.id_app`, `Cat.comp_app` and `Cat.eqToHom_app`
+
+So, for example, if the conclusion of `F` is `f ◁ η = θ` then the conclusion of `F_app` will be
+`η.app (f.obj X) = θ.app X`.
+
+This is useful for automatically generating lemmas that can be applied to expressions of 1-morphisms
+in `Cat` which contain components of 2-morphisms.
+
+There is also a term elaborator `to_app_of% t` for use within proofs.
+-/
+
+open Lean Meta Elab Tactic
+open Mathlib.Tactic
+open Qq
+
+namespace CategoryTheory
+
+/-- Simplify an expression in `Cat` using basic properties of `NatTrans.app`. -/
+def catAppSimp (e : Expr) : MetaM Simp.Result :=
+  simpOnlyNames [
+    ``Cat.whiskerLeft_app, ``Cat.whiskerRight_app, ``Cat.id_app, ``Cat.comp_app,
+    ``Cat.eqToHom_app] e
+    (config := { decide := false })
+
+/--
+Given a term of type `∀ ..., η = θ`, where `η θ : f ⟶ g` are 2-morphisms in some bicategory
+`B`, which is bound by the `∀` binder, get the corresponding equation in the bicategory `Cat`.
+
+It is important here that the levels in the term are level metavariables, as otherwise these will
+not be reassignable to the corresponding levels of `Cat`. -/
+def toCatExpr (e : Expr) : MetaM Expr := do
+  let (args, binderInfos, conclusion) ← forallMetaTelescope (← inferType e)
+  -- Find the expression corresponding to the bicategory, by anylizing `η = θ` (i.e. conclusion)
+  let B ←
+    match conclusion.getAppFnArgs with
+    | (`Eq, #[_, η, _]) =>
+      match (← inferType η).getAppFnArgs with
+      | (`Quiver.Hom, #[_, _, f, _]) =>
+        match (← inferType f).getAppFnArgs with
+        | (`Quiver.Hom, #[_, _, a, _]) =>
+        inferType a
+        | _ => throwError "The conclusion {conclusion} is not an equality of 2-morphisms!"
+      | _ => throwError "The conclusion {conclusion} is not an equality of 2-morphisms!"
+    | _ => throwError "The conclusion {conclusion} is not an equality!"
+  let some BIdx := args.findIdx? (· == B)
+    | throwError "Can not find the bicategory {B} in the arguments this declaration"
+  -- Create level metavariables to be used for `Cat.{v u}`
+  let u ← mkFreshLevelMVar
+  let v ← mkFreshLevelMVar
+  -- Assign `B` to `Cat.{v, u}`
+  let _ ← isDefEq B (.const ``Cat [u, v])
+  -- Assign the right bicategory instance to `Cat.{v, u}`
+  let some inst ← args.findM? fun x => do
+      return (← inferType x).getAppFnArgs == (`CategoryTheory.Bicategory, #[B])
+    | throwError "Can not find the argument for the bicategory instance of the bicategory in which \
+      the equality is taking place."
+  let _ ← isDefEq inst (.const ``CategoryTheory.Cat.bicategory [u, v])
+  -- Construct the new expression
+  let value := mkAppN e args
+  let mvars := (← getMVars value).map (Expr.mvar)
+  -- Erease the binderinfos for the bicategory and the instance
+  let binderInfos := binderInfos.eraseIdx BIdx
+  let binderInfos := binderInfos.eraseIdx BIdx
```

Comment:
It seems that we still need `Bicategory B` immediately after `B`. I attempted to remove this limitation by using `args` instead of `mvars`.
```lean
  let rec
  /-- Recursive function which applies `mkLambdaFVars` stepwise
  (so that each step can have different binderinfos) -/
    apprec (i : Nat) (e : Expr) : MetaM Expr := do
      if i < args.size then
        let arg := args[i]!
        let bi := binderInfos[i]!
        let e' ← apprec (i + 1) e
        unless arg != B && arg != inst do return e'
        mkLambdaFVars #[arg] e' (binderInfoForMVars := bi)
      else
        return e
```

Reply:
Good catch, thanks a lot!